Library/Homebrew/utils/lock.sh

```diff
@@ -0,0 +1,31 @@
+# create a lock using `flock (2)`. A name is required as first argument.
+# the lock will be automatically unlocked when the shell process quits.
+lock() {
+  local name="$1"
+  local lock_file="/tmp/homebrew-${name}.lock"
+  # open the lock file to file descriptor 3, so the shell process can hold the lock.
+  exec 3>"$lock_file"
+  if ! _create_lock
+  then
+    odie <<EOS
+Another active Homebrew process is already in process.
+Please wait for it to finish or terminate it to continue.
+EOS
+  fi
+}
+
+_create_lock() {
+  if [[ -x /usr/bin/ruby ]]
+  then
+    /usr/bin/ruby -e 'exit 1 unless File.new(3).flock(File::LOCK_EX|File::LOCK_NB)'
```

Comment:
Probably a stupid question but colour me curious. Why are we only explicitly checking system tools here instead of more broadly using something like `/usr/bin/env ruby`?